File: s3/R/s3client.R
# Copyright Core DF — Apache License 2.0

s3_missing_env <- function(vars) list(status_code = 601L, error = paste("Environment variables", vars, "should be defined"))
s3_transport_error <- function(message = "inaccessible") list(status_code = 0L, error = message)

`%||%` <- function(a, b) if (is.null(a) || (is.character(a) && !nzchar(a))) b else a

.bucket <- function(explicit) {
  b <- explicit %||% Sys.getenv("S3_BUCKET")
  if (!nzchar(b)) return(NULL)
  b
}

.s3_region <- function() {
  r <- Sys.getenv("AWS_REGION")
  if (nzchar(r)) return(r)
  r <- Sys.getenv("AWS_DEFAULT_REGION")
  if (nzchar(r)) return(r)
  "us-east-1"
}

.s3_endpoint <- function() {
  ep <- Sys.getenv("S3_ENDPOINT_URL")
  if (nzchar(ep)) ep else NULL
}

Init <- function() {
  if (!nzchar(Sys.getenv("AWS_ACCESS_KEY_ID")) && !nzchar(Sys.getenv("AWS_PROFILE"))) {
    return(s3_missing_env("AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY or AWS_PROFILE"))
  }
  if (!nzchar(Sys.getenv("S3_BUCKET"))) {
    return(s3_missing_env("S3_BUCKET (or pass bucket per call)"))
  }
  list(status_code = 200L)
}

GetObject <- function(key, bucket_name = NULL) {
  if (!requireNamespace("aws.s3", quietly = TRUE)) stop("install.packages('aws.s3')")
  b <- .bucket(bucket_name)
  if (is.null(b)) return(s3_missing_env("S3_BUCKET"))
  tryCatch({
    endpoint <- .s3_endpoint()
    txt <- aws.s3::s3read_using(
      function(con) paste(readLines(con, warn = FALSE), collapse = "\n"),
      object = key,
      bucket = b,
      region = .s3_region(),
      endpoint = endpoint
    )
    list(status_code = 200L, content = txt)
  }, error = function(e) s3_transport_error(e$message))
}

PutObject <- function(key, content, bucket_name = NULL) {
  if (!requireNamespace("aws.s3", quietly = TRUE)) stop("install.packages('aws.s3')")
  b <- .bucket(bucket_name)
  if (is.null(b)) return(s3_missing_env("S3_BUCKET"))
  tryCatch({
    endpoint <- .s3_endpoint()
    tf <- tempfile()
    on.exit(unlink(tf), add = TRUE)
    writeLines(content, tf, useBytes = FALSE)
    aws.s3::put_object(
      file = tf,
      object = key,
      bucket = b,
      region = .s3_region(),
      endpoint = endpoint
    )
    list(status_code = 200L)
  }, error = function(e) s3_transport_error(e$message))
}

ListObjects <- function(prefix = "", bucket_name = NULL) {
  if (!requireNamespace("aws.s3", quietly = TRUE)) stop("install.packages('aws.s3')")
  b <- .bucket(bucket_name)
  if (is.null(b)) return(s3_missing_env("S3_BUCKET"))
  tryCatch({
    endpoint <- .s3_endpoint()
    df <- aws.s3::get_bucket_df(
      bucket = b,
      prefix = prefix,
      region = .s3_region(),
      endpoint = endpoint,
      max = 1000L
    )
    keys <- if (is.data.frame(df) && "Key" %in% names(df)) as.character(df$Key) else character(0)
    list(status_code = 200L, keys = keys)
  }, error = function(e) s3_transport_error(e$message))
}

File: s3/Dart/lib/s3client.dart
// Copyright Core DF — Apache License 2.0
import 'dart:convert';
import 'dart:io';

import 'package:minio/minio.dart';

import 'result.dart';

class S3client {
  static String _env(String key, [String fallback = '']) {
    final value = Platform.environment[key] ?? '';
    return value.isEmpty ? fallback : value;
  }

  static String _bucket(String? explicit) {
    if (explicit != null && explicit.isNotEmpty) return explicit;
    return _env('S3_BUCKET');
  }

  static Minio _client() {
    final accessKey = _env('AWS_ACCESS_KEY_ID');
    final secretKey = _env('AWS_SECRET_ACCESS_KEY');
    final region = _env('AWS_REGION', _env('AWS_DEFAULT_REGION', 'us-east-1'));
    final endpoint = _env('S3_ENDPOINT_URL');

    if (endpoint.isNotEmpty) {
      final uri = Uri.parse(endpoint);
      final port = uri.hasPort ? uri.port : (uri.scheme == 'https' ? 443 : 80);
      return Minio(
        endPoint: uri.host,
        port: port,
        accessKey: accessKey,
        secretKey: secretKey,
        useSSL: uri.scheme == 'https',
        region: region,
      );
    }

    return Minio(
      endPoint: 's3.$region.amazonaws.com',
      accessKey: accessKey,
      secretKey: secretKey,
      useSSL: true,
      region: region,
    );
  }

  static Map<String, dynamic> Init() {
    if (_env('AWS_ACCESS_KEY_ID').isEmpty && _env('AWS_PROFILE').isEmpty) {
      return CoreautoResult.missingEnv('AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY or AWS_PROFILE');
    }
    if (_env('S3_BUCKET').isEmpty) {
      return CoreautoResult.missingEnv('S3_BUCKET (or pass bucket per call)');
    }
    return {'status_code': 200};
  }

  static Future<Map<String, dynamic>> GetObject(String key, [String? bucketName]) async {
    final b = _bucket(bucketName);
    if (b.isEmpty) return CoreautoResult.missingEnv('S3_BUCKET');
    try {
      final stream = await _client().getObject(b, key);
      final chunks = <int>[];
      await for (final chunk in stream) {
        chunks.addAll(chunk);
      }
      final content = utf8.decode(chunks, allowMalformed: true);
      return {'status_code': 200, 'content': content};
    } catch (e) {
      return CoreautoResult.transportError(e.toString());
    }
  }

  static Future<Map<String, dynamic>> PutObject(String key, String content, [String? bucketName]) async {
    final b = _bucket(bucketName);
    if (b.isEmpty) return CoreautoResult.missingEnv('S3_BUCKET');
    try {
      await _client().putObject(b, key, Stream.value(utf8.encode(content)), size: utf8.encode(content).length);
      return {'status_code': 200};
    } catch (e) {
      return CoreautoResult.transportError(e.toString());
    }
  }

  static Future<Map<String, dynamic>> ListObjects([String prefix = '', String? bucketName]) async {
    final b = _bucket(bucketName);
    if (b.isEmpty) return CoreautoResult.missingEnv('S3_BUCKET');
    try {
      final keys = <String>[];
      await for (final item in _client().listObjectsV2(b, prefix: prefix)) {
        keys.add(item.key ?? '');
      }
      return {'status_code': 200, 'keys': keys.where((k) => k.isNotEmpty).toList()};
    } catch (e) {
      return CoreautoResult.transportError(e.toString());
    }
  }
}
